receiver/collectdreceiver/collectd.go

```diff
@@ -0,0 +1,292 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collectdreceiver
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/timestamp"
+)
+
+const (
+	collectDMetricDerive   = "derive"
+	collectDMetricGauge    = "gauge"
+	collectDMetricCounter  = "counter"
+	collectDMetricAbsolute = "absolute"
+)
+
+type collectDRecord struct {
+	Dsnames        []*string              `json:"dsnames"`
+	Dstypes        []*string              `json:"dstypes"`
+	Host           *string                `json:"host"`
+	Interval       *float64               `json:"interval"`
+	Plugin         *string                `json:"plugin"`
+	PluginInstance *string                `json:"plugin_instance"`
+	Time           *float64               `json:"time"`
+	TypeS          *string                `json:"type"`
+	TypeInstance   *string                `json:"type_instance"`
+	Values         []*json.Number         `json:"values"`
+	Message        *string                `json:"message"`
+	Meta           map[string]interface{} `json:"meta"`
+	Severity       *string                `json:"severity"`
+}
+
+func (r *collectDRecord) isEvent() bool {
+	return r.Time != nil && r.Severity != nil && r.Message != nil
+}
+
+func (r *collectDRecord) protoTime() *timestamp.Timestamp {
+	if r.Time == nil {
+		return nil
+	}
+	ts := time.Unix(0, int64(float64(time.Second)**r.Time))
+	tsp, err := ptypes.TimestampProto(ts)
+	if err != nil {
+		return nil
+	}
+	return tsp
+}
+
+func (r *collectDRecord) extractAndAppendMetrics(metrics []*metricspb.Metric, defaultLabels map[string]string) ([]*metricspb.Metric, error) {
+	// Ignore if record is an event instead of data point
+	if r.isEvent() {
+		recordEventsReceived()
+		return metrics, nil
+
+	}
+
+	recordMetricsReceived()
+	labels := make(map[string]string, len(defaultLabels))
+	for k, v := range defaultLabels {
+		labels[k] = v
+	}
+
+	for i := range r.Dsnames {
+		if i < len(r.Dstypes) && i < len(r.Values) && r.Values[i] != nil {
+			dsType, dsName, val := r.Dstypes[i], r.Dsnames[i], r.Values[i]
+			metricName, usedDsName := r.getReasonableMetricName(i, labels)
+
+			addIfNotNullOrEmpty(labels, "plugin", true, r.Plugin)
+			parseAndAddLabels(labels, r.PluginInstance, r.Host)
+			addIfNotNullOrEmpty(labels, "dsname", !usedDsName, dsName)
```

Comment:
This is the only place that the 3rd parameter is not the constant true, it seems more readable to have the conditional here and remove the `cond` parameter